### PR TITLE
[Linter] Checks for out of bounds array indexing with a constant index.

### PR DESCRIPTION
**Pull request by tx-tomcat**
## Description 

This PR introduces a new lint for detecting potential out-of-bounds array (vector) indexing in Sui Move code. 
## Implementation Details

The lint operates on the typed AST (Abstract Syntax Tree) of Sui Move code. It focuses on vector operations, particularly `vector::borrow` and `vector::borrow_mut` calls. Here's how it works:

1. **Vector Size Tracking**: The lint maintains a map of vector variables and their known sizes. This is updated when vectors are created or modified (e.g., through `vector::push_back` or `vector::pop_back`).

2. **Index Analysis**: For each vector access, the lint examines the index used. It attempts to determine if the index is a constant value or can be statically analyzed.

3. **Bounds Checking**: The lint compares the index against the known size of the vector. If the index is determined to be out of bounds, a warning is issued.

4. **Dynamic Index Handling**: For cases where the index is determined at runtime, the lint applies heuristics to assess the likelihood of an out-of-bounds access.

## Test plan
Added more use case including true positive, true negative, false positive, false negative case

## Release notes

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: external-crates/move/crates/move-compiler/tests/linter/out_of_bounds_indexing.move [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module 0x42::M {
+    use std::vector;
+
+    fun func1() {
+        let arr = vector[1, 2, 3, 4, 5];
+        vector::borrow(&arr, 10);
+    }
+}
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/tests/linter/out_of_bounds_indexing.move:8]: This should have some tests that use index syntax, plus use the vector mutably.

In addition, we need a test for this:

```
    fun func1() {
        let mut arr = vector[1, 2, 3, 4, 5];
        arr.push(6);
        arr.push(7);
        arr.push(8);
        arr.push(9);
        arr.push(10);
        arr.push(11);
        vector::borrow(&arr, 10);
    }

```

Comment (jamecornel) [reply external-crates/move/crates/move-compiler/tests/linter/out_of_bounds_indexing.move:23]: added more testcase and refactored

**Review by tnowacki — APPROVED** [review]

**Review by tnowacki — CHANGES_REQUESTED** [review]
Sorry for first accepting. After looking deeper at the implementation, I don't think this is a direction for this lint that we can land. 

A version of this that would be land-able would need to use the abstract interpreter. But I don't think that our abstract interpreter has quite enough power to it to do this sort of analysis, since we don't have a mechanism for widening for loops. 

TLDR: We will need to shelve this for now
